src/heuristic/HeuristicMapper.cpp

```diff
@@ -395,23 +400,46 @@ void HeuristicMapper::mapToMinDistance(const std::uint16_t source,
 }
 
 HeuristicMapper::Node HeuristicMapper::aStarMap(size_t layer) {
-  std::vector<std::uint16_t> consideredQubits{};
-  Node                       node{};
+  std::set<std::uint16_t> consideredQubits{};
+  Node                    node{};
+  TwoQubitMultiplicity    twoQubitGateMultiplicity{};
 
-  mapUnmappedGates(layer, consideredQubits);
+  for (const auto& gate : layers.at(layer)) {
+    if (!gate.singleQubit()) {
+      const bool  reverse = gate.control >= gate.target;
+      const auto& q1 =
+          reverse ? gate.target : static_cast<std::uint16_t>(gate.control);
+      const auto& q2 =
+          reverse ? static_cast<std::uint16_t>(gate.control) : gate.target;
+      consideredQubits.emplace(q1);
+      consideredQubits.emplace(q2);
+      if (twoQubitGateMultiplicity.find({q1, q2}) ==
+          twoQubitGateMultiplicity.end()) {
+        twoQubitGateMultiplicity[{q1, q2}] = {!reverse, reverse};
+      } else {
+        if (reverse) {
+          twoQubitGateMultiplicity[{q1, q2}].second++;
+        } else {
+          twoQubitGateMultiplicity[{q1, q2}].first++;
+        }
+      }
+    }
+  }
```

Comment:
Just a suggestion. Probably missing some static_casts.
Might look a little cleaner. Not the biggest fan of ternary if's and the implicit conversion in L418 (although it looks quite elegant). I'll leave it up to you whether you want to adapt.
What I'd do anyway, is to pre-construct the `{q1, q2}` pair in your code.
```suggestion
  for (const auto& gate : layers.at(layer)) {
    if (!gate.singleQubit()) {
      consideredQubits.emplace(gate.control);
      consideredQubits.emplace(gate.target);
      if (gate.control >= gate.target) {
        const auto edge = std::pair{gate.target, gate.control};
        if (twoQubitGateMultiplicity.find(edge) == twoQubitGateMultiplicity.end()) {
          twoQubitGateMultiplicity[edge] = {0, 1};
        } else {
          twoQubitGateMultiplicity[edge].second++;
        }
      } else {
        const auto edge = std::pair{gate.control, gate.target};
        if (twoQubitGateMultiplicity.find(edge) == twoQubitGateMultiplicity.end()) {
          twoQubitGateMultiplicity[edge] = {1, 0};
        } else {
          twoQubitGateMultiplicity[edge].first++;
        }
      }
  }
```

Reply:
Much cleaner, thank you!